### PR TITLE
React to MapIdentityApi HTTP endpoints API review feedback

### LINKED ISSUE #50009: MapIdentityApi HTTP endpoints
## Background and Motivation

So far, we have discussed the design of the dotnet APIs related to `MapIdentityApi` in issues like https://github.com/dotnet/aspnetcore/issues/47227#issuecomment-1491142050 and https://github.com/dotnet/aspnetcore/issues/47228#issue-1626092004, but we have not discussed the shape of the HTTP endpoints. This issue is for that purpose.

### Anonymous endpoints

- POST /register
  - Request: application/json
    ```csharp
    class RegisterRequest
    {
        required string Email { get; init; }
        required string Password { get; init; }
    }
    ```
  - Response: 200 OK (Empty)
  - Response: 400 Bad Request application/problem+json
    - Example:
    ```json
    {
        "type": "https://tools.ietf.org/html/rfc9110#section-15.5.1",
        "title": "One or more validation errors occurred.",
        "status": 400,
        "errors": {
            "PasswordRequiresDigit": [
                "Passwords must have at least one digit ('0'-'9')."
            ],
            "PasswordRequiresLower": [
                "Passwords must have at least one lowercase ('a'-'z')."
            ]
        }
    }
    ```
  - Default sign in options:
    - RequireConfirmedEmail = false
    - RequireConfirmedPhoneNumber = false
    - RequireConfirmedAccount = false
  - Default password options:
    - RequiredLength = 6
    - RequiredUniqueChars = 1
    - RequireDigit = true
    - RequireLowercase = true
    - RequireNonAlphanumeric = true
    - RequireUppercase = true
  - Always attempts to send confirmation email linking to /confirmEmail?userId={userId}&code={code} using the IEmailSender registered with DI.

- POST /login?cookieMode={bool}&persistCookies={bool}
  - Request: application/json
    ```csharp
    class LoginRequest
    {
        required string Email { get; init; }
        required string Password { get; init; }
        string? TwoFactorCode { get; init; }
        string? TwoFactorRecoveryCode { get; init; }
    }
    ```
  - Response: 200 OK
    - Access token: application/json
    ```csharp
    class AccessTokenResponse
    {
        [JsonPropertyName("token_type")]
        string TokenType { get; } = "Bearer";

        [JsonPropertyName("access_token")]
        required string AccessToken { get; init; }

        [JsonPropertyName("expires_in")]
        required long ExpiresInSeconds { get; init; }

        [JsonPropertyName("refresh_token")]
        required string RefreshToken { get; init; }
    }
    ```
    - Cookie (Empty):
    ```
    Set-Cookie: .AspNetCore.Identity.Application={cookieValue}; path=/; secure; samesite=lax; httponly
    ```
  - Response: 401 Unauthorized (Empty)
  - Response: 400 Bad Request application/problem+json
  - Default lockout options:
    - AllowedForNewUsers = true
    - DefaultLockoutTimeSpan = TimeSpan.FromMinutes(5)
    - MaxFailedAccessAttempts = 5
  - Yes we're inconsistent with the property capitalization because I tried to emulate oath.
    - https://github.com/dotnet/aspnetcore/issues/49482 tracks fixing this.

- POST /refresh
  - Request: application/json
    ```csharp
    class RefreshRequest
    {
        required string RefreshToken { get; init; }
    }
    ```
  - Response: 200 OK
    - `AccessTokenResponse`: See /login above
  - Response: 401 Unauthorized (Empty)

- GET /confirmEmail?userId={userId}&code={code}&changedEmail={email}
  - Response: 200 OK text/plain "Thank you for confirming your email."
  - Response: 401 Unauthorized (Empty)

- POST /resendConfirmationEmail
  - Request: application/json
    ```csharp
    class ResendEmailRequest
    {
        required string Email { get; init; }
    }
    ```
  - Response: 200 OK (Empty)
    - This is the response whether or not the email exists in the store.

- POST /forgotPassword
  - Request: application/json
    ```csharp
    class ForgotPasswordRequest
    {
        required string Email { get; init; }
    }
    ```
  - Response: 200 OK (Empty)
  - Response: 400 Bad Request application/problem+json

- POST /resetPassword
  - Request: application/json
    ```csharp
    class ResetPasswordRequest
    {
        required string Email { get; init; }
        required string ResetCode { get; init; }
        required string NewPassword { get; init; }
    }
    ```
  - Response: 200 OK (Empty)
  - Response: 400 Bad Request application/problem+json
  - The expiration of the ResetCode is controlled by the PasswordResetTokenProvider which is a DataProtectorTokenProvider which defaults to 1 day.

### Authenticated endpoints (/account prefix)

All of the below endpoints respond with a 401 Unauthorized if the user is not authenticated.

- POST /account/2fa
  - Request: application/json
  ```csharp
  class TwoFactorRequest
  {
      bool? Enable { get; init; }
      string? TwoFactorCode { get; init; }

      bool ResetSharedKey { get; init; }
      bool ResetRecoveryCodes { get; init; }
      bool ForgetMachine { get; init; }
  }
  ```
  - Response: 200 OK application/json
  ```csharp
  class TwoFactorResponse
  {
      required string SharedKey { get; init; }
      required int RecoveryCodesLeft { get; init; }
      string[]? RecoveryCodes { get; init; }
      required bool IsTwoFactorEnabled { get; init; }
      required bool IsMachineRemembered { get; init; }
  }
  ```
  - Response: 400 Bad Request application/problem+json
  - Unlike previous iterations, this response is no longer available via an authenticated GET request. Clients must send a POST with a non-empty JSON body to prevent CSRF attacks.

- GET /account/info
  - Response: 200 OK application/json
  ```csharp
  class InfoResponse 
  {
      required string Email { get; init; }
      required bool IsEmailConfirmed { get; init; }
      required IDictionary<string, string> Claims { get; init; }
  }
  ```

- POST /account/info
  - Request: application/json
  ```csharp
  class InfoRequest
  {
      string? NewEmail { get; init; }
      string? NewPassword { get; init; }
      string? OldPassword { get; init; }
  }
  ```
  - Response: 200 OK
    - `InfoResponse`: See GET /account/info above
  - Response: 400 Bad Request application/problem+json

## Usage Examples

Assuming that the endpoints were added using `app.MapGroup("/identity").MapIdentityApi<MyUser>();`, you could log in and change your email as follows:

```csharp
// Server side

app.MapGroup("/identity").MapIdentityApi<MyUser>();
app.MapGet("/requires-auth", (ClaimsPrincipal user) => $"Hello, {user.Identity?.Name}!").RequireAuthorization();
```

```csharp
// Client side

// The request body is: { "email": "<email>", "password": "<password>" }
await httpClient.PostAsJsonAsync("/identity/register", new { email, password });
// The request body the same to login since we have not yet configured 2fa: { "email": "<email>", "password": "<password>" }
var loginResponse = await httpClient.PostAsJsonAsync("/identity/login", new { email, password });

// loginResponse is similar to the "Access Token Response" defined in the OAuth 2 spec
// {
//   "token_type": "Bearer",
//   "access_token": "<access-token>",
//   "expires_in": 3600
//   "refresh_token": "<refresh-token>",
// }
var loginContent = await loginResponse.Content.ReadFromJsonAsync<AccessTokenResponse>();

httpClient.DefaultRequestHeaders.Authorization = new("Bearer", loginContent.AccessToken);

// Writes "Hello, <email>!"
Console.WriteLine(await httpClient.GetStringAsync("/requires-auth"));

// getInfoResponse include email and claims
// {
//     "email": "<email>",
//     "isEmailConfirmed": false,
//     "claims": {
//         "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier": "<random-id>",
//         "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name": "<email>",
//         "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress": "<email>",
//         "AspNet.Identity.SecurityStamp": "<security stamp>",
//         "amr": "pwd"
//     }
// }
var getInfoResponse = await httpClient.GetFromJsonAsync<InfoResponse>("/identity/acount/info");

// The request body is: { "newEmail": "<new-email>" }
// The response renmains the same as getInfoResponse until it is confirmed.
var postInfoResponse = await httpClient.PostAsJsonAsync("/identity/acount/info", new { newEmail });

await ConfirmEmail(); // Let's pretend this exists.

// Writes "Hello, <email>!" again because our claims have not updated.
Console.WriteLine(await httpClient.GetStringAsync("/requires-auth"));

// getInfoResponse2 demonstrates the store has been updated but not the user claims yet.
// {
//     "email": "<new-email>",
//     "isEmailConfirmed": false,
//     "claims": {
//         "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier": "<random-id>",
//         "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name": "<email>",
//         "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress": "<email>",
//         "AspNet.Identity.SecurityStamp": "<security stamp>",
//         "amr": "pwd"
//     }
// }
var getInfoResponse2 = await httpClient.GetFromJsonAsync<InfoResponse>("/identity/acount/info");

// The request body is: { "refreshtoken": "<refresh-token>" }
var refreshResponse = await httpClient.PostAsJsonAsync("/identity/refresh", new { refreshToken });

// refreshResponse is identical to loginResponse with new tokens
var refreshContent = await refreshResponse.Content.ReadFromJsonAsync<AccessTokenResponse>();
httpClient.DefaultRequestHeaders.Authorization = new("Bearer", refreshContent.AccessToken);

// Writes "Hello, <new-email>!"
Console.WriteLine(await httpClient.GetStringAsync("/requires-auth"));
```

## Alternative Designs

- Stop using `snake_case` for the `AccessTokenResponse` properties. We're not implementing OAuth, so why are we pretending?
  - This is particularly egregious for `refresh_token` where we require it without the `_` for the `RefreshRequest.RefreshToken` property.
  - See #49482
- Expose all the DTOs as public sealed types in the `Microsoft.AspNetCore.Identity.DTO` for easier endpoint filtering and client usage.
  - See #49424
- Use the `/manage` prefix instead of `/account` prefix for authenticated endpoints.
  - This is more similar to Identity UI
  - If people use `MapGroup("/account")` instead of `MapGroup("/identity")` you'd be forced to make requests to endpoints like `/account/account/info` with the current design.
- 

## Risks

Any API we commit to has backwards compatibility risks. We're planning to ship `MapIdentityApi` in some form though, so we just need to try to be as careful and foward looking as possible.

### PR DESCRIPTION
**Pull request by halter73**
Fixes #50009

### CODE REVIEW

File: src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs [hunk 1]
```diff
@@ -90,12 +90,13 @@ public static IEndpointConventionBuilder MapIdentityApi<TUser>(this IEndpointRou
         });
 
         routeGroup.MapPost("/login", async Task<Results<Ok<AccessTokenResponse>, EmptyHttpResult, ProblemHttpResult>>
-            ([FromBody] LoginRequest login, [FromQuery] bool? cookieMode, [FromQuery] bool? persistCookies, [FromServices] IServiceProvider sp) =>
+            ([FromBody] LoginRequest login, [FromQuery] bool? useCookies, [FromQuery] bool? useSessionCookies, [FromServices] IServiceProvider sp) =>
```
Comment (BrennanConroy) [inline src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs:93]:
```suggestion
            ([FromBody] LoginRequest login, [FromQuery] bool? useCookies, [FromQuery] bool? usePersistentCookies, [FromServices] IServiceProvider sp) =>
```
Was session deliberate?

File: src/Identity/test/Identity.FunctionalTests/MapIdentityApiTests.cs [hunk 2]
```diff
@@ -110,10 +110,43 @@ public async Task CanLoginWithBearerToken(string addIdentityMode)
...
+        });
+        using var client = app.GetTestClient();
+
+        await RegisterAsync(client);
+        var loginResponse = await client.PostAsJsonAsync("/identity/login", new { Email, Password });
+
+        loginResponse.EnsureSuccessStatusCode();
+
         var loginContent = await loginResponse.Content.ReadFromJsonAsync<JsonElement>();
         var tokenType = loginContent.GetProperty("token_type").GetString();
```
Comment (BrennanConroy) [inline src/Identity/test/Identity.FunctionalTests/MapIdentityApiTests.cs:146]: Didn't these change?

Comment (halter73) [reply src/Identity/test/Identity.FunctionalTests/MapIdentityApiTests.cs:146]: Yes, but this test is demonstrating that you can get the old `snake_case` response back by configuring `PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower`.

File: src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs (hunk 1 above)
Comment (halter73) [reply src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs:93]: > Was session deliberate?

It was. I'm going to send an email to update the API proposal. The reason is the default behavior before with just `useCookies=true` was persistent and `persistCookies=false` opted out of that so it effectively defaulted to true.

It was a horrible API, but making the simpler `useCookies=true` persist cookies by default still makes sense in my opinion.

**Review by BrennanConroy — APPROVED** [review]

File: src/Shared/BearerToken/DTO/AccessTokenResponse.cs
Comment (Carl-Binneman) [inline src/Shared/BearerToken/DTO/AccessTokenResponse.cs:1]: Is there any way to bring back snake_case? 
This change really hurts me

**Pull request merged**